### PR TITLE
Desktop: Use the react-select library for tag input

### LINKED ISSUE #1535: Feature Request Please add tags list to Windows Desktop
The Android app has a very elegant interface for adding tags to notes. One can select existing tag(s) or create new one. Please add this functionality to the Win Desktop. I often find myself adding duplicate tags using Windows because I forgot I had already added a similar tag. So I end up with many similar tags. 

Thank you.

### PR DESCRIPTION
**Pull request by CalebJohn**
I finally found a stable input that we can use for tags, [react-select](https://react-select.com/home)
It's very flexible which means we can change the behaviour however we like.

One thing I was sure about is deleting tags with backspace.
    Currently the user cannot delete tags by pressing backspace while the cursor is next to an item. I intentionally disabled this to prevent users from accidentally deleting tags. Personally I would prefer to delete using backspace, so I thought I would reach out for opinions and see what others felt would be the most sane behavior.

image

image

edit: fixes #1535

### CODE REVIEW

**Comment by tessus** [comment]
I think deleting tags with backspace is fine. It's not that you delete the tag, but remove it. Thus re-adding it will not be an issue, if a user accidently removes it.

**Comment by laurent22** [comment]
I also think that using backspace for deletion is fine.

It's nice that you found a lib we can use for this. I'll try your pr soon.

**Comment by CalebJohn** [comment]
Okay, I've enabled backspace to delete tags. I anticipate that there will be a few complaints about it, but I agree that the majority should appreciate it.

**Comment by laurent22** [comment]
Looks good overall, just a few things I've noticed:

- Center text vertically between "Add or remove tags" and dropdown list.
- We've lost the ability to press ESC to exit the dialog box
- Tags need to use theme font
- Ideally the cross to delete a tag should be white (theme.color) when the dark theme is used, but if it's too complicated it's fine in black.

If it's too complicated to customise the tag CSS (i.e. if we need to add special CSS files or rules to theme.js) then just ignore the last two comments.

File: ElectronClient/app/gui/MainScreen.jsx [hunk 1]
```diff
@@ -132,16 +132,19 @@ class MainScreenComponent extends React.Component {
 			});
 		} else if (command.name === 'setTags') {
 			const tags = await Tag.tagsByNoteId(command.noteId);
-			const tagTitles = tags.map((a) => { return a.title }).sort();
+			const tagObjects = tags.map((a) => { return {value: a.id, label: a.title }}).sort();
```
Comment (laurent22) [inline ElectronClient/app/gui/MainScreen.jsx]: I think we need a custom sort function here to sort by title. Calling `sort()` on objects gives probably random results.

Comment (laurent22) [reply ElectronClient/app/gui/MainScreen.jsx]: And could you rename it to `noteTags` please? That way it makes it clear that it's only the current tags of the notes.

File: ElectronClient/app/gui/PromptDialog.jsx [hunk 2]
```diff
@@ -155,6 +191,20 @@ class PromptDialog extends React.Component {
 				timeFormat={time.timeFormat()}
 				onChange={(momentObject) => onDateTimeChange(momentObject)}
 			/>
+		} else if (this.props.inputType === 'tags') {
+			inputComp = <CreatableSelect
+				styles={styles.tagList}
+				theme={styles.tagListTheme}
+				ref={input => this.answerInput_ = input}
```
Comment (laurent22) [inline ElectronClient/app/gui/PromptDialog.jsx]: Please use React.createRef() to create references (the other methods are deprecated).

File: ElectronClient/app/gui/MainScreen.jsx (hunk 1 above)
Comment (CalebJohn) [reply ElectronClient/app/gui/MainScreen.jsx]: You are absolutely right about sort, that was an oversight. It clearly wasn't sorting.

**Comment by CalebJohn** [comment]
All done! I didn't think the css tweaks were possible but when I looked again it was pretty easy to do.

**Comment by laurent22** [comment]
It works great now, thanks a lot @CalebJohn!

**Comment by laurent22** [comment]
By the way, it's fine to have multiple commits on a pull request. In the end I always do a "Squash and merge" anyway so there will be only one commit on master.

**Comment by CalebJohn** [comment]
Thanks @laurent22 that's good to know.

**Pull request merged**